### ISSUE TITLE
Can't reinstall ?

### ISSUE THREAD

**Issue by Camille92** [ctubio/Krypto-trading-bot#12]
Hi Carles,

I'm moving to CentOS machines and to the Alibaba servers I was telling you about but I can't get anything to work :S.

Here is a list of the problems:
node-gyp and rebuild don't install automatically, but a npm install -g fixes it.

Then I have problems with stunnel:

It doesn't get installed so I get: after a npm start:

```
> K.js@3.0.0 test-stunnel /root/K
> test -z "$SKIP_STUNNEL$(ps axu | grep stunnel | grep -v grep)" && exec stunnel dist/K-stunnel.conf || :

sh: line 0: exec: stunnel: not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! K.js@3.0.0 test-stunnel: `test -z "$SKIP_STUNNEL$(ps axu | grep stunnel | grep -v grep)" && exec stunnel dist/K-stunnel.conf || :`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the K.js@3.0.0 test-stunnel script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-06-21T18_33_35_898Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! K.js@3.0.0 start: `npm run test-compile && forever start --uid $npm_package_config_config -a -l /dev/null K.js $npm_package_config_config && exec npm run test-stunnel`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the K.js@3.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-06-21T18_33_35_944Z-debug.log
```

If I install it with npm it doesn't change anything, but if I install it with yum I get:

```
> test -z "$SKIP_STUNNEL$(ps axu | grep stunnel | grep -v grep)" && exec stunnel dist/K-stunnel.conf || :

Clients allowed=31999
stunnel 4.56 on x86_64-redhat-linux-gnu platform
Compiled/running with OpenSSL 1.0.1e-fips 11 Feb 2013
Threading:PTHREAD Sockets:POLL,IPv6 SSL:ENGINE,OCSP,FIPS Auth:LIBWRAP
Reading configuration from file dist/K-stunnel.conf
FIPS mode is enabled
Compression not enabled
PRNG seeded successfully
Initializing service [gdax_stream]
Error loading verify certificates from /usr/local/etc/stunnel/fix.gdax.com.pem
error queue: B084002: error:0B084002:x509 certificate routines:X509_load_cert_crl_file:system lib
error queue: 2006D080: error:2006D080:BIO routines:BIO_new_file:no such file
SSL_CTX_load_verify_locations: 2001002: error:02001002:system library:fopen:No such file or directory
Service [gdax_stream]: Failed to initialize SSL context
str_stats: 10 block(s), 1153 data byte(s), 580 control byte(s)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! K.js@3.0.0 test-stunnel: `test -z "$SKIP_STUNNEL$(ps axu | grep stunnel | grep -v grep)" && exec stunnel dist/K-stunnel.conf || :`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the K.js@3.0.0 test-stunnel script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-06-21T18_30_54_750Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! K.js@3.0.0 start: `npm run test-compile && forever start --uid $npm_package_config_config -a -l /dev/null K.js $npm_package_config_config && exec npm run test-stunnel`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the K.js@3.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-06-21T18_30_54_795Z-debug.log
```

Any idea about what I can do ?

Thank you very much,

Camille

**Comment by ctubio** [comment]
just curious, how did you install stunnel in centos?

to install gdax cert:
` sudo rm -rf /usr/local/etc/stunnel && sudo mkdir -p /usr/local/etc/stunnel/ && openssl s_client -showcerts -connect fix.gdax.com:4198 < /dev/null | openssl x509 -outform PEM > fix.gdax.com.pem && sudo mv fix.gdax.com.pem /usr/local/etc/stunnel/` this was supposed to be run by npm install

**Comment by Camille92** [comment]
thank you!

well just a basic "sudo yum install stunnel"  :)

**Comment by ctubio** [comment]
umf well donno, the npm install was supposed also to include `sudo yum -y install gcc-c++ automake autoconf libtool libxml2 libxml2-devel zlib-devel openssl stunnel`

**Comment by Camille92** [comment]
Oh yeah i forgot to tell you I struggled a bit with those ones before so I ran a sudo yum group install "Development Tools" to fix it that I found there :)

https://www.cyberciti.biz/faq/centos-rhel-7-redhat-linux-install-gcc-compiler-development-tools/

**Comment by ctubio** [comment]
niiice, thanks for the info'¡

**Comment by Camille92** [comment]
Well I think I'll note manage to get it run on CentOS!

Day 2 of trying and I'm blocked there:

```
npm WARN lifecycle K.js@3.0.0~preinstall: cannot run in wd %s %s (wd=%s) K.js@3.0.0 sudo rm -rf /usr/local/etc/stunnel && sudo mkdir -p /usr/local/etc/stunnel/ && exec npm run install-packages /root/K

> node-quickfix-wrap@2.0.14 install /root/K/node_modules/node-quickfix-wrap
> node-gyp rebuild

make: Entering directory `/root/K/node_modules/node-quickfix-wrap/build'
  CXX(target) Release/obj.target/node_quickfix/src/FixSession.o
In file included from ../src/FixSession.cpp:8:0:
../src/FixSession.h:15:30: fatal error: quickfix/Session.h: No such file or directory
 #include "quickfix/Session.h"
                              ^
compilation terminated.
make: *** [Release/obj.target/node_quickfix/src/FixSession.o] Error 1
make: Leaving directory `/root/K/node_modules/node-quickfix-wrap/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:125:13)
gyp ERR! stack     at ChildProcess.emit (events.js:213:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:197:12)
gyp ERR! System Linux 3.10.0-514.10.2.el7.x86_64
gyp ERR! command "/usr/local/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /root/K/node_modules/node-quickfix-wrap
gyp ERR! node -v v8.0.0
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok 
npm ERR! Linux 3.10.0-514.10.2.el7.x86_64
npm ERR! argv "/usr/local/bin/node" "/usr/bin/npm" "install"
npm ERR! node v8.0.0
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE

npm ERR! node-quickfix-wrap@2.0.14 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-quickfix-wrap@2.0.14 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the node-quickfix-wrap package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs node-quickfix-wrap
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls node-quickfix-wrap
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /root/K/npm-debug.log

```

**Comment by Camille92** [comment]
I moved to a Ubuntu distrib but I still can't get it to work. Did you try to do an install recently and did it work ?
I feel like something is broken somewhere :/

**Comment by ctubio** [comment]
the issue now is that quickfix lib is not installed, mostly possible the file /usr/local/lib/libquickfix.so doesnt exist, to create it is supposed that `npm run compile-quickfix` does it

**Comment by ctubio** [comment]
oh sorry i didnt read before about Ubuntu, i use Debian and works fine (like at dummy Docker and super Travis-ci)

**Comment by Camille92** [comment]
Well I tried centOS but as it's working on AWS on Ubuntu I thought, well better keeping on something I know.

But I keep bumping on errors, I tried to run npm run compile-quickfix:

and it ends up like that:
```

quickfix-v.1.14.4/test/test.txt
+ uname
+ libtoolize --copy
./bootstrap: 9: ./bootstrap: libtoolize: not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! K.js@3.0.0 compile-quickfix: `mkdir -p build && cd build && (test -f /usr/local/lib/libquickfix.so || test -f /usr/local/lib/libquickfix.dylib) || (rm -rf quickfix-v* *.tar.gz* && wget https://github.com/quickfix/quickfix/archive/$npm_package_config_quickfix.tar.gz && tar -zxvf v*.tar.gz && cd quickfix-v* && ./bootstrap && ./configure && make && sudo make install && sudo cp config.h /usr/local/include/quickfix/ && (test -n "$(command -v ldconfig)" && sudo ldconfig || :))`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the K.js@3.0.0 compile-quickfix script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-06-23T11_22_03_872Z-debug.log
```

:( That's very annoying haha

**Comment by ctubio** [comment]
`sudo apt-get install libtool automake` should fix it

**Comment by Camille92** [comment]
Yes I don't know either, but it's something with gyp / rebuild and quickfix those are the ones always throwing errors !

Now I go just a step further!
```

quickfix-v.1.14.4/test/test.txt
+ uname
+ libtoolize --copy
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'config'.
libtoolize: copying file 'config/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
+ autoreconf -i -Wall,no-obsolete
./bootstrap: 11: ./bootstrap: autoreconf: not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! K.js@3.0.0 compile-quickfix: `mkdir -p build && cd build && (test -f /usr/local/lib/libquickfix.so || test -f /usr/local/lib/libquickfix.dylib) || (rm -rf quickfix-v* *.tar.gz* && wget https://github.com/quickfix/quickfix/archive/$npm_package_config_quickfix.tar.gz && tar -zxvf v*.tar.gz && cd quickfix-v* && ./bootstrap && ./configure && make && sudo make install && sudo cp config.h /usr/local/include/quickfix/ && (test -n "$(command -v ldconfig)" && sudo ldconfig || :))`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the K.js@3.0.0 compile-quickfix script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-06-23T12_54_07_381Z-debug.log
```

**Comment by ctubio** [comment]
ok now thats a bug :dancer: i missed autoconf , can you run `sudo apt-get install autoconf`? i will add it to the scripts

**Comment by Camille92** [comment]
Well I'm still having erros, I'm going to reset the instance and try again on a 100% clean one to see how it goes. I keep you updated :) !

Hopefully autoconf fixes everything

**Comment by ctubio** [comment]
a clean one for me means to remove `path/to/K` and `/usr/local/lib/libquickfix*` files, then clone and npm start as usual

**Comment by Camille92** [comment]
Yep, same issue:

```
  CXX(target) Release/obj.target/node_quickfix/src/FixAcceptor.o
In file included from ../src/FixAcceptor.h:14:0,
                 from ../src/FixAcceptor.cpp:9:
../src/FixConnection.h:23:29: fatal error: quickfix/config.h: No such file or directory
compilation terminated.
node_quickfix.target.mk:110: recipe for target 'Release/obj.target/node_quickfix/src/FixAcceptor.o' failed
make: *** [Release/obj.target/node_quickfix/src/FixAcceptor.o] Error 1
make: Leaving directory '/K/node_modules/node-quickfix-wrap/build'
```

**Comment by ctubio** [comment]
you are missing the last step of cd build && `rm -rf quickfix-v* *.tar.gz* && wget https://github.com/quickfix/quickfix/archive/$npm_package_config_quickfix.tar.gz && tar -zxvf v*.tar.gz && cd quickfix-v* && ./bootstrap && ./configure && make && sudo make install && sudo cp config.h /usr/local/include/quickfix/`

**Comment by ctubio** [comment]
but is quickfix installed at all? does /usr/local/lib/libquickfix.so exists?

**Comment by Camille92** [comment]
Okay so I ran npm run compile-quickfix that actually went all the way this time!

After if I run npm install I get now:

```
npm WARN lifecycle K.js@3.0.0~preinstall: cannot run in wd %s %s (wd=%s) K.js@3.0.0 sudo rm -rf /usr/local/etc/stunnel && sudo mkdir -p /usr/local/etc/stunnel/ && exec npm run install-packages /K

> node-quickfix-wrap@2.0.14 install /K/node_modules/node-quickfix-wrap
> node-gyp rebuild

make: Entering directory '/K/node_modules/node-quickfix-wrap/build'
  CXX(target) Release/obj.target/node_quickfix/src/FixSession.o
  CXX(target) Release/obj.target/node_quickfix/src/FixLoginProvider.o
  CXX(target) Release/obj.target/node_quickfix/src/FixLoginResponse.o
../src/FixLoginResponse.cpp: In static member function ‘static v8::Handle<v8::Object> FixLoginResponse::wrapFixLoginResponse(FixLoginResponse*)’:
../src/FixLoginResponse.cpp:41:24: warning: variable ‘proto’ set but not used [-Wunused-but-set-variable]
  Local<ObjectTemplate> proto = ctor->PrototypeTemplate();
                        ^
  CXX(target) Release/obj.target/node_quickfix/src/FixAcceptor.o
  CXX(target) Release/obj.target/node_quickfix/src/FixAcceptorStartWorker.o
  CXX(target) Release/obj.target/node_quickfix/src/FixAcceptorStopWorker.o
  CXX(target) Release/obj.target/node_quickfix/src/FixApplication.o
  CXX(target) Release/obj.target/node_quickfix/src/FixConnection.o
  CXX(target) Release/obj.target/node_quickfix/src/FixInitiator.o
../src/FixInitiator.cpp: In static member function ‘static void FixInitiator::sendAsync(const Nan::FunctionCallbackInfo<v8::Value>&, FIX::Message*)’:
../src/FixInitiator.cpp:172:16: warning: unused variable ‘instance’ [-Wunused-variable]
  FixInitiator* instance = Nan::ObjectWrap::Unwrap<FixInitiator>(info.This());
                ^
  CXX(target) Release/obj.target/node_quickfix/src/FixInitiatorStartWorker.o
  CXX(target) Release/obj.target/node_quickfix/src/FixInitiatorStopWorker.o
  CXX(target) Release/obj.target/node_quickfix/src/FixSendWorker.o
  CXX(target) Release/obj.target/node_quickfix/src/node_quickfix.o
  SOLINK(target) Release/obj.target/node_quickfix.node
/usr/bin/ld: cannot find -lxml2
/usr/bin/ld: cannot find -lz
collect2: error: ld returned 1 exit status
node_quickfix.target.mk:145: recipe for target 'Release/obj.target/node_quickfix.node' failed
make: *** [Release/obj.target/node_quickfix.node] Error 1
make: Leaving directory '/K/node_modules/node-quickfix-wrap/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:258:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:194:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
gyp ERR! System Linux 4.4.0-79-generic
gyp ERR! command "/usr/bin/nodejs" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /K/node_modules/node-quickfix-wrap
gyp ERR! node -v v7.10.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok 
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-quickfix-wrap@2.0.14 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-quickfix-wrap@2.0.14 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-06-23T14_02_06_140Z-debug.log
```


But what's is interesting (and maybe the source of the problem) is that when I run:

`cd build && rm -rf quickfix-v* *.tar.gz* && wget https://github.com/quickfix/quickfix/archive/$npm_package_config_quickfix.tar.gz && tar -zxvf v*.tar.gz && cd quickfix-v* && ./bootstrap && ./configure && make && sudo make install && sudo cp config.h /usr/local/include/quickfix/`
What I get is:

Resolving github.com (github.com)... 192.30.255.113, 192.30.255.112
Connecting to github.com (github.com)|192.30.255.113|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2017-06-23 22:05:17 ERROR 404: Not Found.



So basically I tried to look at the link on my computer and indeed : https://github.com/quickfix/quickfix/archive/$npm_package_config_quickfix.tar.gz gives back a 404 !!

**Comment by Camille92** [comment]
OKKKKAAYYYYY IT WORKKKSSSSS!!!!!

https://giphy.com/gifs/happy-fathers-day-fathersday-YEL7FJP6ed008?utm_source=media-link&utm_medium=landing&utm_campaign=Media%20Links&utm_term=https://www.google.com/

What was happening I think it that I did not have the quick fix dependencies installed!

Also I think that as the dependencies are differents from RedHat/CentOS and Ubuntu that might be why the srcipt was not working well -> https://github.com/quickfix/quickfix

**Comment by ctubio** [comment]
im sorry that command i pasted has npm vars ($npm_package_config_quickfix), so your error was expected

**Comment by ctubio** [comment]
we are using this install steps in Debian and centOS (and macOS) as in https://github.com/quickfix/quickfix
